joshprzybyszewski/cribbage
# Redux state management upgrade

## What broke / What you're adding
Addressed issue #66 

## How you did it
The redux state was implemented extremely naively the first time around (@ me). I've made a _HUGE_ upgrade to the client's project structure here.

1. I used [redux toolkit](https://redux-toolkit.js.org/). This massively reduces the amount of boilerplate code needed to create action types and reducers by providing `createSlice`. You can read more about it in their documentation, but it essentially cuts up the redux state into slices used by different parts of the application. As such, it also addresses the issue of "one saga/reducer to rule them all" by using [redux injectors](https://github.com/react-boilerplate/redux-injectors) to dynamically inject them at the site of use. Also noteworthy: redux provides a [great style guide](https://redux.js.org/style-guide/style-guide) that's worth following. The reducers written using `createSlice` _LOOK_ like they're violating the style guide by mutating state, but not to worry - `createSlice` calls `createReducer`, which in turn uses [immer](https://immerjs.github.io/immer/docs/introduction) under the hood to handle copying and updating the state appropriately.

2. I borrowed a _lot_ of ideas from [react boilerplate](https://github.com/react-boilerplate/react-boilerplate). They also use redux toolkit, but unlike redux toolkit, they've used sagas instead of thunks (just like we were doing already). I've pretty much copied the example boilerplate app's code for configuring the redux store, although they were using typescript. A couple other things they do:
    - Use `useDispatch` instead of `connect` and `mapDispatchToProps`. This cleans up the interface to each component by removing the props being mapped, making it clearer when consuming and developing a component where the props are coming from and what is being passed as a prop.
    - Use [selectors](https://github.com/reduxjs/reselect). This will make our application more performant, as it only looks at the part of the state tree we care about at the time of use. It also replaces `connect` and `mapStateToProps`, so a similar benefit as above is achieved. Finally, selectors can create computed state, meaning only the bare minimum needs to actually live in the redux store. A useful example for us is the `selectLoggedIn` selector - it computes a boolean based on the user's id to determine if someone is logged in. We could, of course, just do the same check wherever needed in the code but this is more DRY and reduces cognitive load when reading the code.

3. Also among ideas borrowed from redux toolkit (although the idea itself is broader than that) was the [duck pattern](https://github.com/erikras/ducks-modular-redux). In this pattern, sagas, slices, and subcomponents live together in a directory. It means that finding the state and side effects relevant to a component is very easy. Additionally, I've implemented the "component/container" pattern: any react component which knows about the redux store is deemed a "container," while any component that is purely presentational stays a "component." We don't have any "components" at the moment.

4. I tore out `ConnectedRouter` as suggested [here](https://reacttraining.com/react-router/web/guides/deep-redux-integration) by the react router docs. The router's state shouldn't be conflated with the redux store.

That should be most of the learning summed up. I may have missed some small pieces here and there.

## How to test it and how to try to break it
- In one terminal: `make serve`
- In a second terminal: `make client`
- If it doesn't open your browser, navigate to `localhost:3000`

Play around with the app. It should work exactly as it did before (complete with crappy alert system, still). I haven't added any tests to the frontend yet (to address #64) because I didn't want this already large PR to get too massive.
